Adds test files inclusion for Voyager verification

ametel01 opened the pull request:
Terminal
Adds --test-files flag to include test files in Voyager verification.

This change introduces a new `--test-files` flag to the sncast verify command. When used with the Voyager verifier, this flag includes files in the src/ directory, even those containing "test" in their path, in the source code uploaded for verification.

Changes

- Added a `test_files` boolean field to the Verify struct in `crates/sncast/src/starknet_commands/verify/mod.rs` with a default value of false.
- Modified the verify function in `crates/sncast/src/starknet_commands/verify/mod.rs` to pass the `test_files` flag to the verifier implementations.
- Modified the `VerificationInterface` trait in `crates/sncast/src/starknet_commands/verify/explorer.rs` to include a test_files parameter in the verify method.
- Modified the `Voyager::verify` implementation in `crates/sncast/src/starknet_commands/verify/voyager.rs` to accept the `test_files` flag and pass it to `self.gather_files`.
- Modified the `Voyager::gather_files` function to pass the `include_test_files` flag to the `package_source_files` function.
- Modified the `package_source_files` function to include or exclude test files based on the `include_test_files` flag. Added logic to walk through the path and check for "test" in the lowercase representation.
- Added a new test contract `map_with_tests` including source and test files under `crates/sncast/tests/data/contracts/`.
- Added e2e tests to verify if the `--test-files` flag works correctly by using the new test contract.

Impact

- The `sncast verify` command now accepts a `--test-files flag`.
- When the `--test-files` flag is used with the Voyager verifier, test files under `src/` are included in the verification process.
- If the flag is not present, test files are excluded.
- Walnut verifier remains unaffected by the `test_files` flag.

ametel01 committed:
Adds test files inclusion for Voyager verification

Introduces a `--test-files` flag to the `verify` command.

This flag allows users to include test files located under the `src/` directory during contract verification with the Voyager verifier.

This ensures that all relevant source code, including test-related code, is considered during the verification process, providing a more comprehensive and accurate verification result.

ametel01 committed:
chore: cleanup tests for e2e

ddoktorski commented on crates/sncast/tests/data/contracts/map_with_tests/src/lib.cairo:
Could we simply extend the existing `map` package with these test files instead of adding a new package?

ddoktorski commented on crates/sncast/tests/e2e/verify/voyager.rs:
```diff
@@ -515,3 +515,131 @@ async fn test_error_when_neither_network_nor_url_provided() {
 }
+
+#[tokio::test]
+async fn test_test_files_flag_includes_test_files() {
```
Would it be possible to somehow assert that test files are indeed included? For example by checking the output?

ametel01 replied:
added assertion that the test files are actually included

franciszekjob commented on crates/sncast/src/starknet_commands/verify/walnut.rs:
```diff
@@ -35,6 +35,7 @@ impl VerificationInterface<'_> for WalnutVerificationInterface {
         identifier: ContractIdentifier,
         contract_name: String,
         _package: Option<String>,
+        _test_files: bool,
```
I think we should display throw an error if `--test-files` is used, so users will explicitly see that it doesn't work with walnut verifier (not only in help).

ametel01 replied:
Disallows `--test-files` flag for Walnut verifier and added a test in `e2e::verifier::walnut`

ametel01 committed:
Merge branch 'master' into verifier-test-filter-proc-macros

ametel01 committed:
Adds assertions to verify Voyager files

Extends Voyager tests to assert that the presence or absence
of test files in the request to Voyager is correct.
This ensures that the test files flag functions as expected.

ametel01 committed:
Refactors contract directory structure

Updates the Voyager tests to reflect changes in the location
of the contract files. Simplifies the directory structure by
removing the redundant `map_with_tests` directory and using
`map` instead. This change improves code maintainability
and test clarity.

ametel01 committed:
Disallows `--test-files` flag for Walnut verifier

The `--test-files` flag is not supported when using the Walnut verifier.
This change throws an error when a user attempts to use the flag with the Walnut verifier to prevent unexpected behavior.

ddoktorski commented on crates/sncast/src/starknet_commands/verify/voyager.rs:
```diff
@@ -134,14 +137,22 @@ fn package_source_files(package_metadata: &PackageMetadata) -> Result<Vec<Utf8Pa
+                    }
+                    // Otherwise, skip files with "test" in their path under src/
+                    let path_str = f.path().to_string_lossy().to_lowercase();
+                    if path_str.contains("/test") || path_str.contains("\\test") {
```
What about files with names like `*_test.cairo`?

ddoktorski commented on crates/sncast/src/starknet_commands/verify/walnut.rs:
```diff
@@ -35,8 +35,14 @@ impl VerificationInterface<'_> for WalnutVerificationInterface {
     ) -> Result<VerifyResponse> {
+        if test_files {
+            return Err(anyhow::anyhow!(
+                "The --test-files option is not supported by the walnut verifier"
```
nit:
```suggestion
                "The `--test-files` option is not supported by the walnut verifier"
```

ametel01 committed:
Update crates/sncast/src/starknet_commands/verify/walnut.rs

Co-authored-by: ddoktorski <45050160+ddoktorski@users.noreply.github.com>

ametel01 committed:
skip files ending with _test

ametel01 committed:
Merge branch 'master' into verifier-test-filter-proc-macros

Commit:
fixed typo on test assertion

ametel01 committed:
Merge branch 'master' into verifier-test-filter-proc-macros

ametel01 committed:
Merge branch 'master' into verifier-test-filter-proc-macros

ametel01 commented on crates/sncast/src/starknet_commands/verify/voyager.rs:
```diff
@@ -134,14 +137,26 @@ fn package_source_files(package_metadata: &PackageMetadata) -> Result<Vec<Utf8Pa
+                    if include_test_files {
+                        return true;
+                    }
+                    // Otherwise, skip files with "test" in their path under src/
+                    let path_str = f.path().to_string_lossy().to_lowercase();
+                    if path_str.contains("/test") || path_str.contains("\\test") {
+                        return false;
+                    }
+                    // Also skip files ending with "_test.cairo"
+                    if path_str.ends_with("_test.cairo") {
+                        return false;
+                    }
```
correction, changing the logic make the tests fail:

The current logic specifically targets:
    - Files in test directories: /test or \test in the path
    - Actual test files: files ending with _test.cairo

ametel01 committed:
Merge branch 'master' into verifier-test-filter-proc-macros

franciszekjob reviewed (commented):
Let's mention the new flag in changelog

ametel01 committed:
addressed pr review

ametel01 committed:
updated changelog

ametel01 committed:
Merge branch 'master' into verifier-test-filter-proc-macros

ametel01 commented:
> Let's mention the new flag in changelog

updated

ametel01 committed:
Merge branch 'master' into verifier-test-filter-proc-macros

ametel01 commented on crates/sncast/tests/e2e/verify/walnut.rs:
```diff
@@ -495,17 +509,23 @@ async fn test_test_files_flag_not_supported() {
+
+        {}
+        ",
+            "",
```
add a blank line between the warning and success messages for better readability in the test output, but it was probably unnecessary, removed.

ametel01 committed:
addressed PR review

ddoktorski commented on CHANGELOG.md:
```diff
@@ -47,6 +47,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 #### Added
 
 - `ready` option for `--type` flag in `account create` and `account import` commands (Argent wallet has rebranded as Ready)
+- `--test-files` flag to `verify` command to include test files under src/ for verification (only applies to voyager)
```
Let's move it under the `## [Unreleased]` section

ddoktorski commented on crates/sncast/tests/e2e/verify/voyager.rs:
```diff
@@ -515,3 +515,159 @@ async fn test_error_when_neither_network_nor_url_provided() {
+            "package_name": "map"
+        })))
+        .and(|req: &Request| {
+            if let Ok(body_str) = std::str::from_utf8(&req.body) {
+                if let Ok(body_json) = serde_json::from_str::<serde_json::Value>(body_str) {
+                    if let Some(files) = body_json.get("files") {
+                        if let Some(files_obj) = files.as_object() {
```
Can we replace it with the `let chains` feature, which was stabilized in Rust `1.88.0`? Also in other tests

ametel01 committed:
addressed PR review comments

ametel01 committed:
Merge branch 'master' into verifier-test-filter-proc-macros

ametel01 committed:
fixed changelog section
